File: android/src/main/kotlin/com/eggybyte/content/PluginLogger.kt
package com.eggybyte.content

import android.util.Log
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * A utility object for standardized logging within the plugin.
 *
 * Provides methods for different log levels (DEBUG, INFO, WARNING, ERROR)
 * and ensures consistent formatting including timestamps, log levels, class names,
 * and optional structured details or exception information.
 */
object PluginLogger {

    private const val TAG_PREFIX = "EggyByteSDK"
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())

    /**
     * Represents the different log levels supported by the logger.
     */
    private enum class LogType {
        DEBUG, INFO, WARNING, ERROR
    }

    /**
     * Retrieves the current system timestamp formatted as a string.
     * @return A string representation of the current timestamp.
     */
    private fun getCurrentTimestamp(): String {
        return dateFormat.format(Date())
    }

    /**
     * Formats a log message with a standard structure.
     *
     * @param logType The [LogType] of the message (e.g., DEBUG, INFO).
     * @param className The name of the class from which the log originated.
     * @param message The main log message string.
     * @param details An optional map of key-value pairs providing additional context to the log message.
     * @return The formatted log string.
     */
    private fun formatMessage(logType: LogType, className: String, message: String, details: Map<String, Any?>? = null): String {
        val time = getCurrentTimestamp()
        val detailsString = details?.takeIf { it.isNotEmpty() }?.entries?.joinToString(separator = ", ", prefix = " {", postfix = "}") {
            "${it.key}: '${it.value}'"
        } ?: ""
        return "$time [${logType.name}] ($className): $message$detailsString"
    }

    /**
     * Formats an error message, including optional throwable information and context.
     *
     * @param className The name of the class from which the error originated.
     * @param message The main error message string.
     * @param throwable An optional [Throwable] associated with the error.
     * @param context An optional map of key-value pairs providing additional context to the error.
     * @return The formatted error log string.
     */
    private fun formatErrorMessage(className: String, message: String, throwable: Throwable? = null, context: Map<String, Any?>? = null): String {
        val time = getCurrentTimestamp()
        val contextString = context?.takeIf { it.isNotEmpty() }?.entries?.joinToString(separator = ", ", prefix = " [Context: ", postfix = "]") {
            "${it.key}: '${it.value}'"
        } ?: ""
        val throwableMessage = throwable?.message?.let { " | Exception: $it" } ?: ""
        return "$time [ERROR] ($className): $message$contextString$throwableMessage"
    }

    /**
     * Logs a DEBUG level message.
     *
     * @param className The name of the class from which the log originated.
     * @param message The main log message string.
     * @param details An optional map of key-value pairs providing additional context.
     */
    fun d(className: String, message: String, details: Map<String, Any?>? = null) {
        Log.d("$TAG_PREFIX-$className", formatMessage(LogType.DEBUG, className, message, details))
    }

    /**
     * Logs an INFO level message.
     *
     * @param className The name of the class from which the log originated.
     * @param message The main log message string.
     * @param details An optional map of key-value pairs providing additional context.
     */
    fun i(className: String, message: String, details: Map<String, Any?>? = null) {
        Log.i("$TAG_PREFIX-$className", formatMessage(LogType.INFO, className, message, details))
    }

    /**
     * Logs a WARNING level message.
     *
     * @param className The name of the class from which the log originated.
     * @param message The main log message string.
     * @param details An optional map of key-value pairs providing additional context.
     * @param throwable An optional [Throwable] associated with the warning.
     */
    fun w(className: String, message: String, details: Map<String, Any?>? = null, throwable: Throwable? = null) {
        if (throwable != null) {
            Log.w("$TAG_PREFIX-$className", formatErrorMessage(className, message, throwable, details), throwable)
        } else {
            Log.w("$TAG_PREFIX-$className", formatMessage(LogType.WARNING, className, message, details))
        }
    }

    /**
     * Logs an ERROR level message.
     *
     * @param className The name of the class from which the log originated.
     * @param message The main error message string.
     * @param throwable An optional [Throwable] associated with the error. Its stack trace will be printed.
     * @param context An optional map of key-value pairs providing additional context to the error.
     */
    fun e(className: String, message: String, throwable: Throwable? = null, context: Map<String, Any?>? = null) {
        Log.e("$TAG_PREFIX-$className", formatErrorMessage(className, message, throwable, context), throwable)
    }
} 

File: ios/Classes/PluginLogger.swift
import Foundation
import os.log

/**
 * Unified logging utility for the eggybyte_content plugin.
 *
 * This class provides a centralized logging mechanism with different log levels,
 * consistent formatting, and the ability to filter logs based on build configuration.
 * All logging throughout the plugin should use this class to ensure consistency
 * and easier debugging.
 */
public class PluginLogger {
    
    /// The subsystem identifier for the plugin logs
    private static let subsystem = "com.eggybyte.content"
    
    /// Logger instances for different categories
    private static let generalLogger = OSLog(subsystem: subsystem, category: "General")
    private static let sdkLogger = OSLog(subsystem: subsystem, category: "SDK")
    private static let platformViewLogger = OSLog(subsystem: subsystem, category: "PlatformView")
    private static let communicationLogger = OSLog(subsystem: subsystem, category: "Communication")
    
    /// Plugin tag for consistent log formatting
    private static let pluginTag = "[EggybyteContent]"
    
    // MARK: - General Logging Methods
    
    /**
     * Logs a debug message.
     *
     * Debug messages are only visible in debug builds and provide detailed
     * information for development and troubleshooting.
     *
     * @param message The message to log.
     * @param category The log category (default: general).
     * @param functionName The calling function name (automatically captured).
     * @param fileName The calling file name (automatically captured).
     * @param lineNumber The calling line number (automatically captured).
     */
    public static func debug(
        _ message: String,
        category: LogCategory = .general,
        functionName: String = #function,
        fileName: String = #file,
        lineNumber: Int = #line
    ) {
        #if DEBUG
        let logger = getLogger(for: category)
        let formattedMessage = formatMessage(message, level: "DEBUG", functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        
        // Log to console
        printToConsole(formattedMessage, level: "DEBUG")
        
        // Log to system logger
        os_log("%{public}@", log: logger, type: .debug, formattedMessage)
        #endif
    }
    
    /**
     * Logs an informational message.
     *
     * Info messages provide general information about plugin operations
     * and are visible in both debug and release builds.
     *
     * @param message The message to log.
     * @param category The log category (default: general).
     * @param functionName The calling function name (automatically captured).
     * @param fileName The calling file name (automatically captured).
     * @param lineNumber The calling line number (automatically captured).
     */
    public static func info(
        _ message: String,
        category: LogCategory = .general,
        functionName: String = #function,
        fileName: String = #file,
        lineNumber: Int = #line
    ) {
        let logger = getLogger(for: category)
        let formattedMessage = formatMessage(message, level: "INFO", functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        
        // Log to console
        printToConsole(formattedMessage, level: "INFO")
        
        // Log to system logger
        os_log("%{public}@", log: logger, type: .info, formattedMessage)
    }
    
    /**
     * Logs a warning message.
     *
     * Warning messages indicate potential issues that don't prevent
     * operation but should be addressed.
     *
     * @param message The message to log.
     * @param category The log category (default: general).
     * @param functionName The calling function name (automatically captured).
     * @param fileName The calling file name (automatically captured).
     * @param lineNumber The calling line number (automatically captured).
     */
    public static func warning(
        _ message: String,
        category: LogCategory = .general,
        functionName: String = #function,
        fileName: String = #file,
        lineNumber: Int = #line
    ) {
        let logger = getLogger(for: category)
        let formattedMessage = formatMessage(message, level: "WARNING", functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        
        // Log to console
        printToConsole(formattedMessage, level: "WARNING")
        
        // Log to system logger
        os_log("%{public}@", log: logger, type: .default, formattedMessage)
    }
    
    /**
     * Logs an error message.
     *
     * Error messages indicate serious issues that affect plugin functionality.
     *
     * @param message The message to log.
     * @param error Optional error object for additional context.
     * @param category The log category (default: general).
     * @param functionName The calling function name (automatically captured).
     * @param fileName The calling file name (automatically captured).
     * @param lineNumber The calling line number (automatically captured).
     */
    public static func error(
        _ message: String,
        error: Error? = nil,
        category: LogCategory = .general,
        functionName: String = #function,
        fileName: String = #file,
        lineNumber: Int = #line
    ) {
        let logger = getLogger(for: category)
        var fullMessage = message
        if let error = error {
            fullMessage += " | Error: \(error.localizedDescription)"
        }
        let formattedMessage = formatMessage(fullMessage, level: "ERROR", functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        
        // Log to console
        printToConsole(formattedMessage, level: "ERROR")
        
        // Log to system logger
        os_log("%{public}@", log: logger, type: .error, formattedMessage)
    }
    
    // MARK: - Category-Specific Logging Methods
    
    /**
     * Logs SDK-related messages.
     */
    public static func sdk(_ message: String, level: LogLevel = .info) {
        logWithLevel(message, level: level, category: .sdk)
    }
    
    /**
     * Logs platform view related messages.
     */
    public static func platformView(_ message: String, level: LogLevel = .info) {
        logWithLevel(message, level: level, category: .platformView)
    }
    
    /**
     * Logs communication-related messages.
     */
    public static func communication(_ message: String, level: LogLevel = .info) {
        logWithLevel(message, level: level, category: .communication)
    }
    
    // MARK: - Private Helper Methods
    
    /**
     * Prints a message to the console with timestamp.
     */
    private static func printToConsole(_ message: String, level: String) {
        let timestamp = getCurrentTimestamp()
        let consoleMessage = "\(timestamp) [\(level)] \(message)"
        print(consoleMessage)
        
        // Force flush the console output to ensure immediate display
        fflush(stdout)
    }
    
    /**
     * Gets the current timestamp in HH:mm:ss.SSS format.
     */
    private static func getCurrentTimestamp() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss.SSS"
        return formatter.string(from: Date())
    }
    
    /**
     * Formats a log message with consistent structure.
     */
    private static func formatMessage(
        _ message: String,
        level: String,
        functionName: String,
        fileName: String,
        lineNumber: Int
    ) -> String {
        let shortFileName = (fileName as NSString).lastPathComponent
        return "\(pluginTag) \(shortFileName):\(lineNumber) \(functionName) - \(message)"
    }
    
    /**
     * Gets the appropriate logger for a category.
     */
    private static func getLogger(for category: LogCategory) -> OSLog {
        switch category {
        case .general:
            return generalLogger
        case .sdk:
            return sdkLogger
        case .platformView:
            return platformViewLogger
        case .communication:
            return communicationLogger
        }
    }
    
    /**
     * Logs a message with specified level and category.
     */
    private static func logWithLevel(
        _ message: String,
        level: LogLevel,
        category: LogCategory,
        functionName: String = #function,
        fileName: String = #file,
        lineNumber: Int = #line
    ) {
        switch level {
        case .debug:
            debug(message, category: category, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        case .info:
            info(message, category: category, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        case .warning:
            warning(message, category: category, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        case .error:
            error(message, category: category, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
        }
    }
}

// MARK: - Supporting Enums

/**
 * Log categories for organizing different types of log messages.
 */
public enum LogCategory {
    case general
    case sdk
    case platformView
    case communication
}

/**
 * Log levels for controlling message importance and visibility.
 */
public enum LogLevel {
    case debug
    case info
    case warning
    case error
} 